src/inference.py

```diff
@@ -0,0 +1,159 @@
+import argparse
+
+import torch
+import torch.utils.benchmark as benchmark
+
+from src import log
+from src.log import print_requirements
+
+logger = log.logger
+
+ARCHITECTURES = {
+    "resnet50": "resnet50",
+    "convnext": "convnext_base",
+    "vgg16": "vgg16",
+    "efficient_net_v2": "efficientnet_v2_m",
+    "mobilenet_v3": "mobilenet_v3_large",
+    "resnext50": "resnext50_32x4d",
+    "swin": "swin_b",
+    "vit": "vit_b_16",
+    "ssd_vgg16": "ssd300_vgg16",
+    "fasterrcnn_resnet50_v2": "fasterrcnn_resnet50_fpn_v2",
+}
+
+
+def benchmark_inference(
+    stmt: str,
+    setup: str,
+    input: torch.Tensor,
+    n_runs=100,
+    num_threads: int = 1,
+):
+    """
+    Benchmark a model using torch.utils.benchmark.
+
+    When evaluating model speed in MP/s only the video height, width and batch size are taken into
+    account. The number of channels and sequence length are ignored. Speed evaluation measures
+    how fast can we process an arbitrary input video so channels and sequence length don't
+    affect the model computation speed.
+    """
+
+    timer = benchmark.Timer(
+        stmt=stmt,
+        setup=setup,
+        num_threads=num_threads,
+        globals={"x": input},
+    )
+
+    logger.info(
+        f"Running benchmark on sample of {n_runs} runs with {num_threads} thread(s)..."
+    )
+    result = timer.timeit(n_runs)
+
+    batch, height, width = input.size(0), input.size(-2), input.size(-1)
+    total_pixels = batch * width * height
+
+    logger.info(f"Batch size: {batch}")
+    logger.info(f"Input resolution: {width}x{height} pixels\n")
+
+    mean_per_batch = result.mean
+    median_per_batch = result.median
+
+    mean_speed_mpx = (total_pixels / 1e6) / mean_per_batch
+    median_speed_mpx = (total_pixels / 1e6) / median_per_batch
+
+    logger.info(
+        f"Mean throughoutput per {batch} {width}x{height} px frames: {mean_per_batch:.4f} s"
+    )
+    logger.info(
+        f"Median throughoutput per {batch} {width}x{height} px frames: {median_per_batch:.4f} s\n"
+    )
+
+    logger.info(
+        f"Model mean throughoutput in megapixels per second: {mean_speed_mpx:.3f} MP/s"
+    )
+    logger.info(
+        f"Model median throughoutput in megapixels per second: {median_speed_mpx:.3f} MP/s\n"
+    )
+
+
+def main(args):
+    if args.list_requirements:
+        print_requirements()
+
+    if args.model.lower() not in ARCHITECTURES:
+        raise ValueError("Architecture not supported.")
+
+    stmt = """ \
+    with torch.inference_mode():
+        out = model(x)
+        out = out.clamp(0, 1).float().cpu()
```

Comment:
maybe remove this .clamp since not everyone benchmarking things will be doing image2image problems ?